abi-versions/v0.2.1/README.md

```diff
@@ -0,0 +1,1807 @@
+
+Its arguments (of `arguments_size`) can be retrieved using
+[`proxy_get_buffer_bytes`] with `buffer_id` set to
+`FOREIGN_FUNCTION_ARGUMENTS`.
+
+
+## Unimplemented WASI functions
+
+Various unimplemented WASI functions that are expected to be present
+when the Wasm module is compiled for the `wasm32-wasi` target.
+
+
+### `wasi_snapshot_preview1.args_sizes_get`
+
+* params:
+  - `i32 (size_t *) return_argc`
+  - `i32 (size_t *) return_argv_buffer_size`
+* returns:
+  - `i32 (`[`wasi_errno_t`]`) errno`
+
+Hosts must write `0` to both output arguments to prevent
+[`wasi_snapshot_preview1.args_get`] from being called.
+
+Returned `errno` value is:
+- `SUCCESS` on success.
+- `FAULT` when `return_argc` and/or `return_argv_buffer_size` point to
+  invalid memory address.
+
+
+### `wasi_snapshot_preview1.args_get`
+
+* params:
+  - `i32 (uint8_t **) return_argv`
+  - `i32 (uint8_t *) return_argv_size`
+* returns:
+  - `i32 (`[`wasi_errno_t`]`) errno`
+
+This function should not be called if zero sizes were returned
+in [`wasi_snapshot_preview1.args_sizes_get`].
+
+Returned `errno` value is:
+- `SUCCESS` on success.
+
+
+### `wasi_snapshot_preview1.proc_exit`
+
+* params:
+  - `i32 (uint32_t) exit_code`
+* returns:
+  - none
+
+This function is never called.
+
+
+# Types
+
+### `proxy_log_level_t`
+
+- `TRACE` = `0`
+- `DEBUG` = `1`
+- `INFO` = `2`
+- `WARN` = `3`
+- `ERROR` = `4`
+- `CRITICAL` = `5`
+
+
+### `proxy_status_t`
+
+- `OK` = `0`
+- `NOT_FOUND` = `1`
+- `BAD_ARGUMENT` = `2`
+- `SERIALIZATION_FAILURE` = `3`
+- `PARSE_FAILURE` = `4`
+- `INVALID_MEMORY_ACCESS` = `6`
+- `EMPTY` = `7`
+- `CAS_MISMATCH` = `8`
+- `INTERNAL_FAILURE` = `10`
+- `UNIMPLEMENTED` = `12`
+
+
+### `proxy_action_t`
+
+- `CONTINUE` = `0`
+- `PAUSE` = `1`
+
+
+### `proxy_buffer_type_t`
+
+- `HTTP_REQUEST_BODY` = `0`
+- `HTTP_RESPONSE_BODY` = `1`
+- `DOWNSTREAM_DATA` = `2`
+- `UPSTREAM_DATA` = `3`
+- `HTTP_CALL_RESPONSE_BODY` = `4`
+- `GRPC_CALL_MESSAGE` = `5`
+- `VM_CONFIGURATION` = `6`
+- `PLUGIN_CONFIGURATION` = `7`
+- `FOREIGN_FUNCTION_ARGUMENTS` = `8`
+
+
+### `proxy_header_map_type_t`
+
+- `HTTP_REQUEST_HEADERS` = `0`
+- `HTTP_REQUEST_TRAILERS` = `1`
+- `HTTP_RESPONSE_HEADERS` = `2`
+- `HTTP_RESPONSE_TRAILERS` = `3`
+- `GRPC_CALL_INITIAL_METADATA` = `4`
+- `GRPC_CALL_TRAILING_METADATA` = `5`
+- `HTTP_CALL_RESPONSE_HEADERS` = `6`
+- `HTTP_CALL_RESPONSE_TRAILERS` = `7`
+
+
+### `proxy_peer_type_t`
+
+- `UNKNOWN` = `0`
+- `LOCAL` = `1`
+- `REMOTE` =`2`
+
+
+### `proxy_stream_type_t`
+
+- `HTTP_REQUEST` = `0`
+- `HTTP_RESPONSE` = `1`
+- `DOWNSTREAM` = `2`
+- `UPSTREAM` = `3`
+
+
+### `proxy_metric_type_t`
+
+- `COUNTER` = `0`
+- `GAUGE` = `1`
+- `HISTOGRAM` = `2`
+
+
+### `wasi_errno_t`
+
+- `SUCCESS` = `0`
+- `BADF` = `8`
+- `FAULT` = `21`
+- `INVAL` = `28`
+- `NOTSUP` = `58`
+
+
+### `wasi_fd_id_t`
+
+- `STDOUT` = `1`
+- `STDERR` = `2`
+
+
+### `wasi_clock_id_t`
+
+- `REALTIME` = `0`
+- `MONOTONIC` = `1`
+
+
+# Serialization
+
+## HTTP fields and gRPC metadata
+
+TODO
```

Comment:
Data encoding is almost as important as the function declarations.

Reply:
Right, I'll add this shortly. I left `TODO` because I didn't want to delay this PR any longer, but I intend to complete this sections before merging this PR.